tuf/src/interchange/cjson/shims.rs

```diff
@@ -368,82 +369,109 @@ pub struct PublicKeyValue {
             .cloned()
             .collect::<Vec<metadata::TargetPath>>();
         paths.sort();
-        let mut key_ids = meta
+
+        let mut key_ids = delegation
             .key_ids()
             .iter()
             .cloned()
             .collect::<Vec<crypto::KeyId>>();
         key_ids.sort();
 
         Delegation {
-            role: meta.role().clone(),
-            terminating: meta.terminating(),
-            threshold: meta.threshold(),
+            name: delegation.name().clone(),
+            terminating: delegation.terminating(),
+            threshold: delegation.threshold(),
             key_ids,
             paths,
         }
     }
+}
 
-    pub fn try_into(self) -> Result<metadata::Delegation> {
-        let paths = self
-            .paths
-            .iter()
-            .cloned()
-            .collect::<HashSet<metadata::TargetPath>>();
-        if paths.len() != self.paths.len() {
-            return Err(Error::Encoding("Non-unique delegation paths.".into()));
-        }
+impl TryFrom<Delegation> for metadata::Delegation {
+    type Error = Error;
 
-        let key_ids = self
-            .key_ids
-            .iter()
-            .cloned()
-            .collect::<HashSet<crypto::KeyId>>();
-        if key_ids.len() != self.key_ids.len() {
+    fn try_from(delegation: Delegation) -> Result<Self> {
+        let delegation_key_ids_len = delegation.key_ids.len();
+        let key_ids = delegation.key_ids.into_iter().collect::<HashSet<_>>();
+
+        if key_ids.len() != delegation_key_ids_len {
             return Err(Error::Encoding("Non-unique delegation key IDs.".into()));
         }
 
-        metadata::Delegation::new(self.role, self.terminating, self.threshold, key_ids, paths)
+        let delegation_paths_len = delegation.paths.len();
+        let paths = delegation.paths.into_iter().collect::<HashSet<_>>();
+
+        if paths.len() != delegation_paths_len {
+            return Err(Error::Encoding("Non-unique delegation paths.".into()));
+        }
+
+        metadata::Delegation::new(
+            delegation.name,
+            delegation.terminating,
+            delegation.threshold,
+            key_ids,
+            paths,
+        )
     }
 }
 
 #[derive(Serialize, Deserialize)]
 pub struct Delegations {
     #[serde(deserialize_with = "deserialize_reject_duplicates::deserialize")]
     keys: BTreeMap<crypto::KeyId, crypto::PublicKey>,
-    roles: Vec<metadata::Delegation>,
+    roles: Vec<Delegation>,
 }
 
-impl Delegations {
-    pub fn from(delegations: &metadata::Delegations) -> Delegations {
+impl From<&metadata::Delegations> for Delegations {
+    fn from(delegations: &metadata::Delegations) -> Delegations {
+        let mut roles = delegations
+            .roles()
+            .iter()
+            .map(Delegation::from)
+            .collect::<Vec<Delegation>>();
+
+        // We want our roles in a consistent order.
+        roles.sort_by(|lhs, rhs| lhs.name.cmp(&rhs.name));
```

Comment:
nit: use `sort_by_key`?

Reply:
I'd love to be able to use `roles.sort_by_key(|r| &r.name)`, but unfortunately that gives this error:

```
449 |         roles.sort_by_key(|role| &role.name);
    |                            ----- ^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                            |   |
    |                            |   return type of closure is &'2 MetadataPath
    |                            has type `&'1 shims::Delegation`
```

Reply:
oh I remember this https://github.com/rust-lang/rust/issues/34162